stm32: fix a few bugs with ADC on STM32WBxx MCUs and tweak tests so they pass on NUCLEO_WB55

dpgeorge opened the pull request:
### Summary

A NUCLEO_WB55 is part of the Octoprobe hardware test stack.  This PR aims to get the test suite fully passing on this board.

The main thing here is to tweak the existing `tests/ports/stm32` tests, which were originally written for pyboards (PYBv1.x and PYBD-SFx).  These tests should now pass on a wider variety of stm32 boards.

### Testing

Tested on a NUCLEO_WB55 via `./run-tests.py -t 0 -d ports/stm32`.

CI and Octoprobe will also test this PR.

dpgeorge committed:
stm32/adc: Apply re-read errata for WB55.

Following 17898f8607dc4fb881e860719cc1906d304e60f4.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
stm32/adc: Simplify ADC calibration settings.

Combine the common settings for L1/L4/WB with existing G0/G4/H5 settings.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
stm32/adc: Fix core temperature reading on WB55.

It needs a divisor of 100 because the calibration temperatures are 30 and
130 degrees, similar to the H5.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
stm32/machine_adc: Fix internal ADC channel reading on WB MCUs.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/extmod/machine_uart_tx.py: Support STM32WB boards.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/ports/stm32: Tweak tests to run on a wider set of boards.

There should be no change to these tests for existing PYBV1x and PYBD_SFx
boards.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
Octoprobe run results are here: https://reports.octoprobe.org/github_selfhosted_testrun_127/octoprobe_summary_report.html

That looks a lot better, although still a few tests are failing.

hmaerki commented:
* I added PYBV11. Testrun 128 in processing (but does not include your latest commit).
* My 2f2c-NUCLEO_WB55 is slow, even when I connect it to my notebook! So I assume it has to to with the board or the firmware update we did together yesterday [flash nucleo-wb55 bluetooth-stack](https://www.octoprobe.org/testbed_micropython/developer_guide/10_debugging.html#dut-nucleo-wb55-remotely-flash-bluetooth-stack)
* These are the jumper settings (according to tentacle_specs.py): https://www.maerki.com/hans/download/damien/tentacle_2f2c-NUCLEO_WB55_2025-06-25a.jpg

@dpgeorge: Which usb connection do you use? I use the one behind 'RESET'.
```
    USB_STLINK CN15: located on the back side of 'LED4'
    USB_MCU    CN1 : located on the back side of 'RESET'
```

dpgeorge committed:
tests/extmod_hardware: Add UART config for STM32WB boards.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/extmod_hardware/machine_uart_irq_rxidle.py: Ignore inital IRQ.

On stm32, the hardware generates an RXIDLE IRQ after enabling the UART,
because the RX line is technically idle.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
I have updated the tests here so that they should all now work on NUCLEO_WB55.

If you bridge PA2/PA3 on the NUCLEO_WB55 (Arduino D0/D1 on CN9, or 35/37 on CN10) then that will allow the UART tests to pass.

hmaerki commented:
> If you bridge PA2/PA3 on the NUCLEO_WB55 (Arduino D0/D1 on CN9, or 35/37 on CN10) then that will allow the UART tests to pass.

Stupid me. Thanks for your help! I didn't read the table in datasheet correctly. I will correct it tomorrow morning and run another test.

dpgeorge committed:
stm32/uart: Suppress additional RX idle IRQs on F4/L1.

These MCUs only clear the RX idle IRQ if the data register is read, which
won't occur if the only IRQ is the RX idle IRQ (because then reading and
discarding the DR may lead to lost data).

To work around this, explicitly suppress the RX idle IRQ so that it's only
passed through to the Python callback once.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/extmod_hardware/machine_uart_irq_rxidle.py: Test multiple writes.

This tests that the RXIDLE callback is called correctly after a second lot
of bytes are received.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/extmod/select_poll_eintr.py: Skip test if target can't bind.

Eg on PYBV10 with THREAD variant, the firmware has both the `_thread` and
`socket` modules but no NIC.

Signed-off-by: Damien George <damien@micropython.org>
